In [1]:
import pandas as pd

In [2]:
companies = pd.read_csv("C:\\Users\\HP\\Desktop\\L3\\remedial\\Spark Funds Industrial Assignment - 1\\companies.txt" , 
                        sep='\t', encoding= "ISO-8859-1")

In [271]:
companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [4]:
companies[(companies["name"] == "Blink")]

permalink   name              homepage_url  \
7449    /Organization/Blink  Blink       http://www.blink.la   
7450  /Organization/Blink-4  Blink   http://www.blinnnk.com/   
7451  /Organization/Blink-7  Blink   http://blinkforhome.com   
7452  /Organization/Blink-9  Blink  http://www.flyblink.com/   

                                          category_list     status  \
7449  Content|Journalism|Location Based Services|Med...  operating   
7450                                           Software  operating   
7451                                        Electronics  operating   
7452          Location Based Services|Service Providers  operating   

     country_code state_code         region       city  founded_at  
7449          USA         NY  New York City   New York  01-01-2013  
7450          JPN         18    JPN - Other      China  01-01-2014  
7451          USA         MA         Boston     Boston         NaN  
7452          GBR         N7      Camberley  Camberley         NaN

In [5]:
companies.isnull().sum()

permalink            0
name                 1
homepage_url      5058
category_list     3148
status               0
country_code      6958
state_code        8547
region            8030
city              8028
founded_at       15221
dtype: int64

In [6]:
companies["name"].value_counts()

Blink                                            4
Peach                                            4
Spire                                            4
Roost                                            4
Karma                                            4
                                                ..
Game Time Giving, Inc                            1
Game Time Tickets                                1
Game Trading technologies, Inc.                  1
Game Trust                                       1
İnovatiff Reklam ve Tanıtım Hizmetleri Tic    1
Name: name, Length: 66102, dtype: int64

In [7]:
companies.nunique()

permalink        66368
name             66102
homepage_url     61191
category_list    27296
status               4
country_code       137
state_code         311
region            1092
city              5111
founded_at        3978
dtype: int64

In [8]:
round2 = pd.read_csv("C:\\Users\HP\\Desktop\\L3\\remedial\\Spark Funds Industrial Assignment - 1\\rounds2.csv", 
                 encoding= "ISO-8859-1")
round2

company_permalink  \
0                                     /organization/-fame   
1                                  /ORGANIZATION/-QOUNTER   
2                                  /organization/-qounter   
3                     /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                                   /organization/0-6-com   
...                                                   ...   
114944                          /organization/zzzzapp-com   
114945                          /ORGANIZATION/ZZZZAPP-COM   
114946                               /organization/ãeron   
114947                             /ORGANIZATION/ÔASYS-2   
114948  /organization/ä°novatiff-reklam-ve-tanä±tä±m-h...   

                                funding_round_permalink funding_round_type  \
0       /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1       /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2       /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3       /funding-round/650b8f704416801069bb178a1418776b            venture   
4       /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
...                                                 ...                ...   
114944  /funding-round/8f6d25b8ee4199e586484d817bceda05   convertible_note   
114945  /funding-round/ff1aa06ed5da186c84f101549035d4ae               seed   
114946  /funding-round/59f4dce44723b794f21ded3daed6e4fe            venture   
114947  /funding-round/35f09d0794651719b02bbfd859ba9ff5               seed   
114948  /funding-round/af942869878d2cd788ef5189b435ebc4              grant   

       funding_round_code   funded_at  raised_amount_usd  
0                       B  05-01-2015         10000000.0  
1                       A  14-10-2014                NaN  
2                     NaN  01-03-2014           700000.0  
3                       B  30-01-2014          3406878.0  
4                       A  19-03-2008          2000000.0  
...                   ...         ...                ...  
114944                NaN  01-03-2014            41313.0  
114945                NaN  01-05-2013            32842.0  
114946                  A  01-08-2014                NaN  
114947                NaN  01-01-2015            18192.0  
114948                NaN  01-10-2013            14851.0  

[114949 rows x 6 columns]

In [9]:
round2.rename(columns={"company_permalink":"permalink"}, inplace=True)

In [10]:
round2.isnull().mean()*100

permalink                   0.000000
funding_round_permalink     0.000000
funding_round_type          0.000000
funding_round_code         72.909725
funded_at                   0.000000
raised_amount_usd          17.390321
dtype: float64

In [11]:
round2.nunique()[0]

90247

In [12]:
# getting a copy of companies data
comp = companies.copy()

In [13]:
# getting a copy of round2 data
r2 = round2.copy()

In [14]:
# removing special character present in companies
comp['permalink'] = comp['permalink'].str.replace('[^a-zA-Z0-9\s]', '')

C:\Users\HP\AppData\Local\Temp/ipykernel_32632/4243429802.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  comp['permalink'] = comp['permalink'].str.replace('[^a-zA-Z0-9\s]', '')


In [15]:
# removing special character present in companies
r2['permalink'] = r2['permalink'].str.replace('[^a-zA-Z0-9\s]', '')

C:\Users\HP\AppData\Local\Temp/ipykernel_32632/1921705648.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  r2['permalink'] = r2['permalink'].str.replace('[^a-zA-Z0-9\s]', '')


In [16]:
# changing the character of permalink present in companies and round2 dataset to lower case
comp['permalink'] = comp['permalink'].str.lower()
r2['permalink'] = r2['permalink'].str.lower()

In [17]:
comp.head(1)

permalink   name        homepage_url category_list     status  \
0  organizationfame  #fame  http://livfame.com         Media  operating   

  country_code state_code  region    city founded_at  
0          IND         16  Mumbai  Mumbai        NaN

In [18]:
r2.head(1)

permalink                          funding_round_permalink  \
0  organizationfame  /funding-round/9a01d05418af9f794eebff7ace91f638   

  funding_round_type funding_round_code   funded_at  raised_amount_usd  
0            venture                  B  05-01-2015         10000000.0

In [19]:
# merging round2 and companies
df = pd.merge(r2,comp, on="permalink")
df.head(1)

permalink                          funding_round_permalink  \
0  organizationfame  /funding-round/9a01d05418af9f794eebff7ace91f638   

  funding_round_type funding_round_code   funded_at  raised_amount_usd   name  \
0            venture                  B  05-01-2015         10000000.0  #fame   

         homepage_url category_list     status country_code state_code  \
0  http://livfame.com         Media  operating          IND         16   

   region    city founded_at  
0  Mumbai  Mumbai        NaN

# Checkpoint 1 : Data Cleaning 1

# Table 1.1

# How many unique companies are present in rounds2?

In [20]:
# Extracting premalink and name feature from merged dataframe and checking for nan values
df1 = df[["permalink","name"]]
df1.isnull().sum()

permalink    0
name         1
dtype: int64

In [21]:
# Due to a single nan value present in name column dropping that
df1.dropna(inplace=True)

C:\Users\HP\anaconda3\lib\site-packages\pandas\util\_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


In [22]:
df1.nunique()[1]

66102

#### Unique Companies present in round2 is 66102

# How many unique companies are present in companies?

In [23]:
# Extracting premalink and name feature from companies dataframe

In [24]:
df2 = companies[["permalink","name"]]

In [25]:
# Due to a single nan value present in name column dropping that
df2.dropna(inplace=True)

In [26]:
# After drop checking for nan
df2.isnull().sum()

permalink    0
name         0
dtype: int64

In [27]:
df2.nunique()[1]

66102

#### Unique Companies present in companies is 66102

# In the companies data frame, which column can be used as the unique key for each company? Write the name of the column.

In [28]:
companies.nunique()

permalink        66368
name             66102
homepage_url     61191
category_list    27296
status               4
country_code       137
state_code         311
region            1092
city              5111
founded_at        3978
dtype: int64

#### In the companies data frame, "company_permalink" can be used as the unique key for each company

# Are there any companies in the rounds2 file which are not  present in companies ? Answer Y/N.

In [29]:
df.nunique()

permalink                   66290
funding_round_permalink    114949
funding_round_type             14
funding_round_code              8
funded_at                    5033
raised_amount_usd           22095
name                        66102
homepage_url                61191
category_list               27296
status                          4
country_code                  137
state_code                    311
region                       1092
city                         5111
founded_at                   3978
dtype: int64

In [30]:
df.head()

permalink  \
0             organizationfame   
1          organizationqounter   
2          organizationqounter   
3  organizationtheoneoftheminc   
4            organization06com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd                    name  \
0                  B  05-01-2015         10000000.0                   #fame   
1                  A  14-10-2014                NaN                :Qounter   
2                NaN  01-03-2014           700000.0                :Qounter   
3                  B  30-01-2014          3406878.0  (THE) ONE of THEM,Inc.   
4                  A  19-03-2008          2000000.0                 0-6.com   

             homepage_url                                      category_list  \
0      http://livfame.com                                              Media   
1  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
2  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
3     http://oneofthem.jp                                  Apps|Games|Mobile   
4      http://www.0-6.com                                        Curated Web   

      status country_code state_code      region           city  founded_at  
0  operating          IND         16      Mumbai         Mumbai         NaN  
1  operating          USA         DE  DE - Other  Delaware City  04-09-2014  
2  operating          USA         DE  DE - Other  Delaware City  04-09-2014  
3  operating          NaN        NaN         NaN            NaN         NaN  
4  operating          CHN         22     Beijing        Beijing  01-01-2007

In [31]:
companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

#### No, There are no any companies in the round2 file which are not present in companies.

# Merge the two data frames so that all  variables (columns)  in the companies frame are added to the rounds2 data frame. Name the merged frame master_frame. How many observations are present in master_frame ?

In [32]:
master_frame =  pd.merge(comp,r2, on="permalink")
master_frame.shape[0]

115200

#### After merging two data frames there are 115200 observations are present in maste_frame

# Checkpoint 2 : Funding Type Analysis

In [33]:
master_frame.head()

permalink                    name  \
0             organizationfame                   #fame   
1          organizationqounter                :Qounter   
2          organizationqounter                :Qounter   
3  organizationtheoneoftheminc  (THE) ONE of THEM,Inc.   
4            organization06com                 0-6.com   

             homepage_url                                      category_list  \
0      http://livfame.com                                              Media   
1  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
2  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
3     http://oneofthem.jp                                  Apps|Games|Mobile   
4      http://www.0-6.com                                        Curated Web   

      status country_code state_code      region           city  founded_at  \
0  operating          IND         16      Mumbai         Mumbai         NaN   
1  operating          USA         DE  DE - Other  Delaware City  04-09-2014   
2  operating          USA         DE  DE - Other  Delaware City  04-09-2014   
3  operating          NaN        NaN         NaN            NaN         NaN   
4  operating          CHN         22     Beijing        Beijing  01-01-2007   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [34]:
master_frame.isnull().mean()

permalink                  0.000000
name                       0.000009
homepage_url               0.053490
category_list              0.029748
status                     0.000000
country_code               0.075538
state_code                 0.095295
region                     0.088524
city                       0.088498
founded_at                 0.178689
funding_round_permalink    0.000000
funding_round_type         0.000000
funding_round_code         0.729019
funded_at                  0.000000
raised_amount_usd          0.173880
dtype: float64

In [35]:
master_frame["funding_round_type"].value_counts()

venture                  55618
seed                     30606
debt_financing            6903
angel                     6104
undisclosed               4904
equity_crowdfunding       3263
private_equity            2287
grant                     2204
convertible_note          1823
post_ipo_equity            638
product_crowdfunding       412
non_equity_assistance      191
post_ipo_debt              152
secondary_market            95
Name: funding_round_type, dtype: int64

In [36]:
#########################################################################################

In [37]:
master_frame["raised_amount_usd"] = master_frame["raised_amount_usd"].dropna()

In [38]:
master_frame

permalink  \
0                                     organizationfame   
1                                  organizationqounter   
2                                  organizationqounter   
3                          organizationtheoneoftheminc   
4                                    organization06com   
...                                                ...   
115195                          organizationzzzzappcom   
115196                          organizationzzzzappcom   
115197                                organizationeron   
115198                               organizationasys2   
115199  organizationnovatiffreklamvetantmhizmetleritic   

                                                 name            homepage_url  \
0                                               #fame      http://livfame.com   
1                                            :Qounter  http://www.qounter.com   
2                                            :Qounter  http://www.qounter.com   
3                              (THE) ONE of THEM,Inc.     http://oneofthem.jp   
4                                             0-6.com      http://www.0-6.com   
...                                               ...                     ...   
115195                          Zzzzapp Wireless ltd.  http://www.zzzzapp.com   
115196                          Zzzzapp Wireless ltd.  http://www.zzzzapp.com   
115197                                         ÁERON    http://www.aeron.hu/   
115198                                         Ôasys    http://www.oasys.io/   
115199  İnovatiff Reklam ve Tanıtım Hizmetleri Tic    http://inovatiff.com   

                                            category_list     status  \
0                                                   Media  operating   
1       Application Platforms|Real Time|Social Network...  operating   
2       Application Platforms|Real Time|Social Network...  operating   
3                                       Apps|Games|Mobile  operating   
4                                             Curated Web  operating   
...                                                   ...        ...   
115195        Advertising|Mobile|Web Development|Wireless  operating   
115196        Advertising|Mobile|Web Development|Wireless  operating   
115197                                                NaN  operating   
115198  Consumer Electronics|Internet of Things|Teleco...  operating   
115199                 Consumer Goods|E-Commerce|Internet  operating   

       country_code state_code       region           city  founded_at  \
0               IND         16       Mumbai         Mumbai         NaN   
1               USA         DE   DE - Other  Delaware City  04-09-2014   
2               USA         DE   DE - Other  Delaware City  04-09-2014   
3               NaN        NaN          NaN            NaN         NaN   
4               CHN         22      Beijing        Beijing  01-01-2007   
...             ...        ...          ...            ...         ...   
115195          HRV         15        Split          Split  13-05-2012   
115196          HRV         15        Split          Split  13-05-2012   
115197          NaN        NaN          NaN            NaN  01-01-2011   
115198          USA         CA  SF Bay Area  San Francisco  01-01-2014   
115199          NaN        NaN          NaN            NaN         NaN   

                                funding_round_permalink funding_round_type  \
0       /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1       /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2       /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3       /funding-round/650b8f704416801069bb178a1418776b            venture   
4       /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
...                                                 ...                ...   
115195  /funding-round/8f6d25b8ee4199e586484d817bceda05   convertible_note 

# Average funding amount of venture type

In [39]:
master_frame.drop(columns=["funding_round_code","status","state_code","region","city","founded_at",
                           "funding_round_code","funded_at"],axis=1,inplace=True)

In [40]:
master_frame.head()

permalink                    name  \
0             organizationfame                   #fame   
1          organizationqounter                :Qounter   
2          organizationqounter                :Qounter   
3  organizationtheoneoftheminc  (THE) ONE of THEM,Inc.   
4            organization06com                 0-6.com   

             homepage_url                                      category_list  \
0      http://livfame.com                                              Media   
1  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
2  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
3     http://oneofthem.jp                                  Apps|Games|Mobile   
4      http://www.0-6.com                                        Curated Web   

  country_code                          funding_round_permalink  \
0          IND  /funding-round/9a01d05418af9f794eebff7ace91f638   
1          USA  /funding-round/22dacff496eb7acb2b901dec1dfe5633   
2          USA  /funding-round/b44fbb94153f6cdef13083530bb48030   
3          NaN  /funding-round/650b8f704416801069bb178a1418776b   
4          CHN  /funding-round/5727accaeaa57461bd22a9bdd945382d   

  funding_round_type  raised_amount_usd  
0            venture         10000000.0  
1            venture                NaN  
2               seed           700000.0  
3            venture          3406878.0  
4            venture          2000000.0

In [41]:
master_frame.drop(columns=["homepage_url"],axis=1,inplace=True)

In [42]:
master_frame.isnull().sum()

permalink                      0
name                           1
category_list               3427
country_code                8702
funding_round_permalink        0
funding_round_type             0
raised_amount_usd          20031
dtype: int64

In [44]:
master_frame["category_list"] = master_frame["category_list"].dropna()

In [45]:
master_frame["country_code"] = master_frame["country_code"].dropna()

In [46]:
master_frame.dropna(inplace=True)

In [47]:
master_frame.isnull().sum()

permalink                  0
name                       0
category_list              0
country_code               0
funding_round_permalink    0
funding_round_type         0
raised_amount_usd          0
dtype: int64

# Average funding amount of venture type

In [48]:
venture = master_frame[(master_frame["funding_round_type"]=="venture")]

In [49]:
round(venture["raised_amount_usd"].mean(),2)

11728248.37

### Venture = 11.72 Million USD

# Average funding amount of angel type

In [50]:
angel = master_frame[(master_frame["funding_round_type"]=="angel")]

In [51]:
round(angel["raised_amount_usd"].mean(),2)

970589.61

### Angel = 0.97 Million USD

# Average funding amount of seed type

In [52]:
seed = master_frame[(master_frame["funding_round_type"]=="seed")]

In [53]:
round(seed["raised_amount_usd"].mean(),2)

747916.88

### Seed = 0.74 Million USD

# Average funding amount of private equity type

In [63]:
private_equity = master_frame[(master_frame["funding_round_type"]=="private_equity")]

In [64]:
round(private_equity["raised_amount_usd"].mean(),2)

73244084.68

### Private equity = 73.24 Million USD

# Considering that Spark Funds wants to invest between 5 to 15 million USD per  investment round, which investment type is the most suitable for them?

#### Considering that Spark Funds wants to invest between 5 to 15 million USD per investment round, so venture type ismost suitable for them beacause investment round is 10.63 million USD which meets the Sparx Funds Criteria

# Checkpoint 3: Country Analysis

### 1. Spark Funds wants to see the top nine countries which have received the highest total funding (across ALL sectors for the chosen investment type)


In [62]:
venture.groupby(by="country_code")["raised_amount_usd"].sum().sort_values(ascending=False).apply(lambda x: '{:.1f}'.format(x)).to_frame().reset_index()

country_code raised_amount_usd
0           USA    420844252957.0
1           CHN     39338918773.0
2           GBR     20177983004.0
3           IND     14306468105.0
4           CAN      9496661165.0
..          ...               ...
93          MCO          657000.0
94          SAU          500000.0
95          CMR          359561.0
96          GTM          300000.0
97          MMR          200000.0

[98 rows x 2 columns]

### 2. For the chosen investment type, make a data frame named top9 with the top nine countries (based on the total investment amount each country has received)

In [94]:
top9 = venture.groupby(by="country_code")["raised_amount_usd"].sum().sort_values(ascending=False).apply(lambda x: '{:.1f}'.format(x)).to_frame().reset_index().head(9)
top9

country_code raised_amount_usd
0          USA    420844252957.0
1          CHN     39338918773.0
2          GBR     20177983004.0
3          IND     14306468105.0
4          CAN      9496661165.0
5          FRA      7226851352.0
6          ISR      6854350477.0
7          DEU      6306921981.0
8          JPN      3173647127.0

# Table 3.1: Analysing the Top 3 English-Speaking Countrie

In [66]:
url = "https://en.wikipedia.org/wiki/List_of_countries_and_territories_where_English_is_an_official_language"

In [67]:
eng = pd.read_html(url, header=0)

In [68]:
ce = eng[0]

In [69]:
ge = eng[1]

In [70]:
merged_df = pd.merge(ce, ge, how='outer')

In [74]:
merged_df.head(pd.merge(ce, ge, how='outer'))

Country ISO code Geographic region  Population1  \
0  Antigua and Barbuda      ATG         Caribbean        85000   
1       The Bahamas[1]      BHS         Caribbean       331000   
2          Barbados[2]      BRB         Caribbean       294000   
3            Belize[3]      BLZ   Central America       288000   
4          Botswana[3]      BWA            Africa      1882000   

                     Primary language?  
0  Yes (English-based creole language)  
1  Yes (English-based creole language)  
2  Yes (English-based creole language)  
3  Yes (English-based creole language)  
4                                  Yes

In [96]:
top9.rename(columns={"country_code":"ISO code"}, inplace=True)

In [191]:
master_frame.rename(columns={"country_code":"ISO code"}, inplace=True)
code = pd.merge(master_frame,merged_df, on="ISO code",how="inner")

In [192]:
code.head()

permalink           name category_list ISO code  \
0                organizationfame          #fame         Media      IND   
1     organization21diamondsindia     21Diamonds    E-Commerce      IND   
2  organization247learningprivate  24x7 Learning        EdTech      IND   
3           organization33coupons      33Coupons      Internet      IND   
4               organization3dsoc          3DSoC            3D      IND   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/6de7ffef8091ba9f33821f4b861f434a            venture   
2  /funding-round/4cd5cd165bc869bc40705a9e4b626bee            venture   
3  /funding-round/fc7f4eba112e0df25e9a56b037657820              angel   
4  /funding-round/bbedf1410fddeed341018583bbd0ca1e            venture   

   raised_amount_usd      Country Geographic region  Population1  \
0         10000000.0  India[3][6]              Asia   1393500000   
1          6369507.0  India[3][6]              Asia   1393500000   
2          4000000.0  India[3][6]              Asia   1393500000   
3            20000.0  India[3][6]              Asia   1393500000   
4          1240000.0  India[3][6]              Asia   1393500000   

                   Primary language?  
0  No (but official and educational)  
1  No (but official and educational)  
2  No (but official and educational)  
3  No (but official and educational)  
4  No (but official and educational)

In [97]:
english = pd.merge(top9, merged_df,how="inner", on="ISO code" )
english

ISO code raised_amount_usd         Country Geographic region  Population1  \
0      USA    420844252957.0   United States     North America    328239523   
1      GBR     20177983004.0  United Kingdom            Europe     66040229   
2      IND     14306468105.0     India[3][6]              Asia   1393500000   
3      CAN      9496661165.0          Canada     North America     38048738   

                                   Primary language?  
0                                                Yes  
1                                                Yes  
2                  No (but official and educational)  
3  Yes (Co-official with French, and a predominan...

In [111]:
top3eng = english[english['Primary language?'].str.startswith('Yes')].head(3).reset_index()
top3eng

index ISO code raised_amount_usd         Country Geographic region  \
0      0      USA    420844252957.0   United States     North America   
1      1      GBR     20177983004.0  United Kingdom            Europe   
2      3      CAN      9496661165.0          Canada     North America   

   Population1                                  Primary language?  
0    328239523                                                Yes  
1     66040229                                                Yes  
2     38048738  Yes (Co-official with French, and a predominan...

# 1. Top English-speaking country

In [115]:
print("Top English-speaking country is : ",top3eng["Country"][0])

Top English-speaking country is :  United States


### United States	

# 2. Second English-speaking country


In [116]:
print("Top English-speaking country is : ",top3eng["Country"][1])

Top English-speaking country is :  United Kingdom


### United Kingdom

# 3. Third English-speaking country

In [117]:
print("Top English-speaking country is : ",top3eng["Country"][2])

Top English-speaking country is :  Canada


### Canada

# Checkpoint 4: Sector Analysis 1

In [142]:
# reading mapping.csv
m = pd.read_csv("C:\\Users\\HP\Desktop\\L3\\remedial\\Spark Funds Industrial Assignment - 1\\mapping.csv")
m.head()

category_list  Automotive & Sports  Blanks  Cleantech / Semiconductors  \
0            NaN                    0       1                           0   
1             3D                    0       0                           0   
2    3D Printing                    0       0                           0   
3  3D Technology                    0       0                           0   
4     Accounting                    0       0                           0   

   Entertainment  Health  Manufacturing  News, Search and Messaging  Others  \
0              0       0              0                           0       0   
1              0       0              1                           0       0   
2              0       0              1                           0       0   
3              0       0              1                           0       0   
4              0       0              0                           0       0   

   Social, Finance, Analytics, Advertising  
0                                        0  
1                                        0  
2                                        0  
3                                        0  
4                                        1

### 1. Extract the primary sector of each category list from the category_list colum

In [143]:
m["category_list"].unique()

array([nan, '3D', '3D Printing', '3D Technology', 'Accounting',
       'Active Lifestyle', 'Ad Targeting', 'Advanced Materials',
       'Adventure Travel', 'Advertising', 'Advertising Exchanges',
       'Advertising Networks', 'Advertising Platforms', 'Advice',
       'Aerospace', 'Agriculture', 'Air Pollution Control', 'Algorithms',
       'All Markets', 'All Students', 'Alter0tive Medicine', 'Alumni',
       'A0lytics', 'Android', 'Angels', 'Animal Feed',
       'Anything Capital Intensive', 'App Discovery', 'App Marketing',
       'App Stores', 'Application Performance Monitoring',
       'Application Platforms', 'Apps', 'Aquaculture', 'Architecture',
       'Archiving', 'Art', 'Artificial Intelligence', 'Artists Globally',
       'Assisitive Technology', 'Assisted Living', 'Auctions', 'Audio',
       'Audiobooks', 'Augmented Reality', 'Auto', 'Automated Kiosk',
       'Automotive', 'B2B', 'B2B Express Delivery', 'Babies',
       'Baby Accessories', 'Baby Boomers', 'Baby Safety', 'B

In [144]:
m["category_list"]=m["category_list"].str.replace("0","na")

In [145]:
m.head()

category_list  Automotive & Sports  Blanks  Cleantech / Semiconductors  \
0            NaN                    0       1                           0   
1             3D                    0       0                           0   
2    3D Printing                    0       0                           0   
3  3D Technology                    0       0                           0   
4     Accounting                    0       0                           0   

   Entertainment  Health  Manufacturing  News, Search and Messaging  Others  \
0              0       0              0                           0       0   
1              0       0              1                           0       0   
2              0       0              1                           0       0   
3              0       0              1                           0       0   
4              0       0              0                           0       0   

   Social, Finance, Analytics, Advertising  
0                                        0  
1                                        0  
2                                        0  
3                                        0  
4                                        1

In [146]:
melt_df = pd.melt(m,id_vars="category_list",value_name="value")
melt_df

category_list                                 variable  value
0                  NaN                      Automotive & Sports      0
1                   3D                      Automotive & Sports      0
2          3D Printing                      Automotive & Sports      0
3        3D Technology                      Automotive & Sports      0
4           Accounting                      Automotive & Sports      0
...                ...                                      ...    ...
6187         Wholesale  Social, Finance, Analytics, Advertising      0
6188  Wine And Spirits  Social, Finance, Analytics, Advertising      0
6189          Wireless  Social, Finance, Analytics, Advertising      0
6190             Women  Social, Finance, Analytics, Advertising      0
6191      Young Adults  Social, Finance, Analytics, Advertising      0

[6192 rows x 3 columns]

In [148]:
melt_df1 = melt_df[melt_df['value']==1]
melt_df1

category_list                                 variable  value
8      Adventure Travel                      Automotive & Sports      1
14            Aerospace                      Automotive & Sports      1
45                 Auto                      Automotive & Sports      1
46      Automated Kiosk                      Automotive & Sports      1
47           Automotive                      Automotive & Sports      1
...                 ...                                      ...    ...
6121  Social Recruiting  Social, Finance, Analytics, Advertising      1
6122  Social Television  Social, Finance, Analytics, Advertising      1
6123      Social Travel  Social, Finance, Analytics, Advertising      1
6134    Stock Exchanges  Social, Finance, Analytics, Advertising      1
6167    Venture Capital  Social, Finance, Analytics, Advertising      1

[688 rows x 3 columns]

### 2. Use the mapping file 'mapping.csv' to map each primary sector to one of the eight main sectors (Note that ‘Others’ is also considered one of the main sectors)

In [149]:
master_frame["category_list"] = master_frame["category_list"].apply(lambda x:x.split('|')[0])

In [151]:
master_frame.head()

permalink                    name  \
0                 organizationfame                   #fame   
2              organizationqounter                :Qounter   
4                organization06com                 0-6.com   
6    organization01gamestechnology      01Games Technology   
7  organization0ndinebiomedicalinc  Ondine Biomedical Inc.   

           category_list country_code  \
0                  Media          IND   
2  Application Platforms          USA   
4            Curated Web          CHN   
6                  Games          HKG   
7          Biotechnology          CAN   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
6  /funding-round/7d53696f2b4f607a2f2a8cbb83d01839        undisclosed   
7  /funding-round/2b9d3ac293d5cdccbecff5c8cb0f327d               seed   

   raised_amount_usd  
0         10000000.0  
2           700000.0  
4          2000000.0  
6            41250.0  
7            43360.0

In [157]:
final = pd.merge(master_frame,melt_df1, on="category_list")
final.head()

permalink            name category_list country_code  \
0         organizationfame           #fame         Media          IND   
1        organization90min           90min         Media          GBR   
2        organization90min           90min         Media          GBR   
3        organization90min           90min         Media          GBR   
4  organizationadanceforme  A Dance for Me         Media          USA   

                           funding_round_permalink   funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638              venture   
1  /funding-round/21a2cbf6f2fb2a1c2a61e04bf930dfe6              venture   
2  /funding-round/bd626ed022f5c66574b1afe234f3c90d              venture   
3  /funding-round/fd4b15e8c97ee2ffc0acccdbe1a98810              venture   
4  /funding-round/9ab9dbd17bf010c79d8415b2c22be6fa  equity_crowdfunding   

   raised_amount_usd       variable  value  
0         10000000.0  Entertainment      1  
1         15000000.0  Entertainment      1  
2          5800000.0  Entertainment      1  
3         18000000.0  Entertainment      1  
4          1090000.0  Entertainment      1

In [160]:
final.shape

(88377, 9)

# Checkpoint 5: Sector Analysis 2

In [168]:
df1 = final[(final['funding_round_type']=='venture')]

In [179]:
code  # english[english['Primary language?'].str.startswith('Yes')].head(3).reset_index()

permalink                     name  \
0                     organizationfame                    #fame   
1          organization21diamondsindia               21Diamonds   
2       organization247learningprivate            24x7 Learning   
3                organization33coupons                33Coupons   
4                    organization3dsoc                    3DSoC   
...                                ...                      ...   
73228            organizationorderbook                Orderbook   
73229     organizationplanneddeparture        Planned Departure   
73230              organizationpornhub                  Pornhub   
73231  organizationstgeorgesuniversity  St. George's University   
73232                organizationunyqe                    Unyqe   

       category_list ISO code  \
0              Media      IND   
1         E-Commerce      IND   
2             EdTech      IND   
3           Internet      IND   
4                 3D      IND   
...              ...      ...   
73228        Bitcoin      DMA   
73229  Digital Media      BLZ   
73230   Sex Industry      BLZ   
73231       Colleges      GRD   
73232           Apps      SYC   

                               funding_round_permalink   funding_round_type  \
0      /funding-round/9a01d05418af9f794eebff7ace91f638              venture   
1      /funding-round/6de7ffef8091ba9f33821f4b861f434a              venture   
2      /funding-round/4cd5cd165bc869bc40705a9e4b626bee              venture   
3      /funding-round/fc7f4eba112e0df25e9a56b037657820                angel   
4      /funding-round/bbedf1410fddeed341018583bbd0ca1e              venture   
...                                                ...                  ...   
73228  /funding-round/51f3fb65bb7a71891c1dc6b26025cfc9                angel   
73229  /funding-round/dbda967ab29515ac2ac191e7266a1b0d  equity_crowdfunding   
73230  /funding-round/3687f5e79f2773922bda9d4c2bc1f53e  equity_crowdfunding   
73231  /funding-round/9c2fb34fc442085634469fada2ddcc13       private_equity   
73232  /funding-round/4bed8ef9d7a4c5e3f9c08dfe38872401                 seed   

       raised_amount_usd        Country      Geographic region  Population1  \
0             10000000.0    India[3][6]                   Asia   1393500000   
1              6369507.0    India[3][6]                   Asia   1393500000   
2              4000000.0    India[3][6]                   Asia   1393500000   
3                20000.0    India[3][6]                   Asia   1393500000   
4              1240000.0    India[3][6]                   Asia   1393500000   
...                  ...            ...                    ...          ...   
73228           100000.0    Dominica[1]              Caribbean        73000   
73229           196615.0      Belize[3]        Central America       288000   
73230             4299.0      Belize[3]        Central America       288000   
73231        750000000.0     Grenada[1]              Caribbean       111000   
73232            35000.0  Seychelles[1]  Africa / Indian Ocean        87000   

                         Primary language?  
0        No (but official and educational)  
1        No (but official and educational)  
2        No (but official and educational)  
3        No (but official and educational)  
4        No (but official and educational)  
...                                    ...  
73228   Yes (French-based creole language)  
73229  Yes (English-based creole language)  
73230  Yes (English-based creole language)  
73231  Yes (English-based creole language)  
73232                                   No  

[73233 rows x 11 columns]

In [204]:
#code1 = code[code['Primary language?'].str.startswith('Yes')]
code["Primary language?"] = code["Primary language?"].apply(lambda x:x.split("(")[0])
code

permalink                     name  \
0                     organizationfame                    #fame   
1          organization21diamondsindia               21Diamonds   
2       organization247learningprivate            24x7 Learning   
3                organization33coupons                33Coupons   
4                    organization3dsoc                    3DSoC   
...                                ...                      ...   
73228            organizationorderbook                Orderbook   
73229     organizationplanneddeparture        Planned Departure   
73230              organizationpornhub                  Pornhub   
73231  organizationstgeorgesuniversity  St. George's University   
73232                organizationunyqe                    Unyqe   

       category_list ISO code  \
0              Media      IND   
1         E-Commerce      IND   
2             EdTech      IND   
3           Internet      IND   
4                 3D      IND   
...              ...      ...   
73228        Bitcoin      DMA   
73229  Digital Media      BLZ   
73230   Sex Industry      BLZ   
73231       Colleges      GRD   
73232           Apps      SYC   

                               funding_round_permalink   funding_round_type  \
0      /funding-round/9a01d05418af9f794eebff7ace91f638              venture   
1      /funding-round/6de7ffef8091ba9f33821f4b861f434a              venture   
2      /funding-round/4cd5cd165bc869bc40705a9e4b626bee              venture   
3      /funding-round/fc7f4eba112e0df25e9a56b037657820                angel   
4      /funding-round/bbedf1410fddeed341018583bbd0ca1e              venture   
...                                                ...                  ...   
73228  /funding-round/51f3fb65bb7a71891c1dc6b26025cfc9                angel   
73229  /funding-round/dbda967ab29515ac2ac191e7266a1b0d  equity_crowdfunding   
73230  /funding-round/3687f5e79f2773922bda9d4c2bc1f53e  equity_crowdfunding   
73231  /funding-round/9c2fb34fc442085634469fada2ddcc13       private_equity   
73232  /funding-round/4bed8ef9d7a4c5e3f9c08dfe38872401                 seed   

       raised_amount_usd        Country      Geographic region  Population1  \
0             10000000.0    India[3][6]                   Asia   1393500000   
1              6369507.0    India[3][6]                   Asia   1393500000   
2              4000000.0    India[3][6]                   Asia   1393500000   
3                20000.0    India[3][6]                   Asia   1393500000   
4              1240000.0    India[3][6]                   Asia   1393500000   
...                  ...            ...                    ...          ...   
73228           100000.0    Dominica[1]              Caribbean        73000   
73229           196615.0      Belize[3]        Central America       288000   
73230             4299.0      Belize[3]        Central America       288000   
73231        750000000.0     Grenada[1]              Caribbean       111000   
73232            35000.0  Seychelles[1]  Africa / Indian Ocean        87000   

      Primary language?  
0                   No   
1                   No   
2                   No   
3                   No   
4                   No   
...                 ...  
73228              Yes   
73229              Yes   
73230              Yes   
73231              Yes   
73232                No  

[73233 rows x 11 columns]

In [215]:
c1 = code[(code["ISO code"]=="USA")]
c2 = code[(code["ISO code"]=="GBR")]
c3 = code[(code["ISO code"]=="CAN")]

In [206]:
code.to_csv("C:\\Users\\HP\\Desktop\\L3\\remedial\\Spark Funds Industrial Assignment - 1\\stack_fund.csv")

### 1. Total number of investments (count)

In [223]:
print("Total number of investment in country 1 :",c1.shape[0])

Total number of investment in country 1 : 61791


In [224]:
print("Total number of investment in country 2 :",c2.shape[0])

Total number of investment in country 2 : 5004


In [225]:
print("Total number of investment in country 3 :",c3.shape[0])

Total number of investment in country 3 : 2603


### 2. Total amount of investment (USD))

In [230]:
print("Total amount of investment in  country 1 : ",round(c1["raised_amount_usd"].sum()/1000000,2),"M$")

Total amount of investment in  country 1 :  666635.3 M$


In [231]:
print("Total amount of investment in  country 2 : ",round(c2["raised_amount_usd"].sum()/1000000,2),"M$")

Total amount of investment in  country 2 :  32543.56 M$


In [232]:
print("Total amount of investment in  country 3 : ",round(c3["raised_amount_usd"].sum()/1000000,2),"M$")

Total amount of investment in  country 3 :  18319.08 M$


### 3. Top sector (based on count of investments)

In [238]:
c1.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[:1]

category_list
Biotechnology    8005
Name: raised_amount_usd, dtype: int64

In [239]:
c2.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[:1]

category_list
Biotechnology    414
Name: raised_amount_usd, dtype: int64

In [240]:
c3.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[:1]

category_list
Biotechnology    275
Name: raised_amount_usd, dtype: int64

### 4. Second-best sector (based on count of investments)

In [243]:
c1.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[1:2]

category_list
Software    4411
Name: raised_amount_usd, dtype: int64

In [244]:
c2.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[1:2]

category_list
Software    328
Name: raised_amount_usd, dtype: int64

In [245]:
c3.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[1:2]

category_list
Software    185
Name: raised_amount_usd, dtype: int64

### 5. Third-best sector (based on count of investments)

In [246]:
c1.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[2:3]

category_list
Advertising    2790
Name: raised_amount_usd, dtype: int64

In [247]:
c2.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[2:3]

category_list
E-Commerce    247
Name: raised_amount_usd, dtype: int64

In [248]:
c3.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[2:3]

category_list
Clean Technology    115
Name: raised_amount_usd, dtype: int64

### 6. Number of investments in the top sector (refer to point 3)

In [249]:
c1.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[:1]

category_list
Biotechnology    8005
Name: raised_amount_usd, dtype: int64

In [250]:
c2.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[:1]

category_list
Biotechnology    414
Name: raised_amount_usd, dtype: int64

In [251]:
c3.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[:1]

category_list
Biotechnology    275
Name: raised_amount_usd, dtype: int64

### 7 . Number of investments in the secondbest sector (refer to point 4)

In [252]:
c1.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[1:2]

category_list
Software    4411
Name: raised_amount_usd, dtype: int64

In [253]:
c2.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[1:2]

category_list
Software    328
Name: raised_amount_usd, dtype: int64

In [254]:
c3.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[1:2]

category_list
Software    185
Name: raised_amount_usd, dtype: int64

### 8. Number of investments in the third-best sector (refer to point 5)

In [255]:
c1.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[2:3]

category_list
Advertising    2790
Name: raised_amount_usd, dtype: int64

In [256]:
c2.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[2:3]

category_list
E-Commerce    247
Name: raised_amount_usd, dtype: int64

In [257]:
c3.groupby(by="category_list")["raised_amount_usd"].count().sort_values(ascending=False).iloc[2:3]

category_list
Clean Technology    115
Name: raised_amount_usd, dtype: int64

### 9. For the top sector count-wise (point 3), which company received the highest

In [265]:
c1.groupby(by=["name"])["raised_amount_usd"].sum().sort_values(ascending=False).iloc[:1]

name
Verizon Communications    3.007950e+10
Name: raised_amount_usd, dtype: float64

In [266]:
c2.groupby(by=["name"])["raised_amount_usd"].sum().sort_values(ascending=False).iloc[:1]

name
SoftViaNet    915000000.0
Name: raised_amount_usd, dtype: float64

In [267]:
c3.groupby(by=["name"])["raised_amount_usd"].sum().sort_values(ascending=False).iloc[:1]

name
BlackBerry    1.000000e+09
Name: raised_amount_usd, dtype: float64

### 10. For the second-best sector count-wise (point 4), which company received the highest investment?

In [268]:
c1.groupby(by=["name"])["raised_amount_usd"].sum().sort_values(ascending=False).iloc[1:2]

name
Freescale Semiconductor    1.760000e+10
Name: raised_amount_usd, dtype: float64

In [269]:
c2.groupby(by=["name"])["raised_amount_usd"].sum().sort_values(ascending=False).iloc[1:2]

name
Helios Towers Africa    630000000.0
Name: raised_amount_usd, dtype: float64

In [270]:
c3.groupby(by=["name"])["raised_amount_usd"].sum().sort_values(ascending=False).iloc[1:2]

name
WSP Global    486000000.0
Name: raised_amount_usd, dtype: float64